src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,829 @@
+
+                    actionsBuilder.Add(CreateNewCodeAction(
+                        FeaturesResources.Introduce_overload_with_new_parameter_for_0, document,
+                        expression, methodSymbol, containingMethod, allOccurrences: false, trampoline: false, overload: true, cancellationToken));
+                    actionsBuilder.Add(CreateNewCodeAction(
+                         FeaturesResources.Introduce_overload_with_new_parameter_for_all_occurrences_of_0, document,
+                         expression, methodSymbol, containingMethod, allOccurrences: true, trampoline: false, overload: true, cancellationToken));
+                }
+            }
+
+            return actionsBuilder.ToImmutable();
+
+            // Local function to create a code action with more ease
+            MyCodeAction CreateNewCodeAction(string actionName, Document document,
+                TExpressionSyntax expression, IMethodSymbol methodSymbol, SyntaxNode containingMethod,
+                bool allOccurrences, bool trampoline,
+                bool overload, CancellationToken cancellationToken)
+            {
+                return new MyCodeAction(actionName, c => IntroduceParameterAsync(
+                    document, expression, methodSymbol, containingMethod, allOccurrences, trampoline, overload, cancellationToken));
+            }
+        }
+
+        /// <summary>
+        /// Finds the matches of the expression within the same block.
+        /// </summary>
+        protected static async Task<IEnumerable<TExpressionSyntax>> FindMatchesAsync(Document originalDocument,
+            TExpressionSyntax expressionInOriginal, SyntaxNode withinNodeInCurrent,
+            bool allOccurrences, CancellationToken cancellationToken)
+        {
+            var syntaxFacts = originalDocument.GetRequiredLanguageService<ISyntaxFactsService>();
+            var originalSemanticModel = await originalDocument.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var result = new HashSet<TExpressionSyntax>();
+            var matches = from nodeInCurrent in withinNodeInCurrent.DescendantNodesAndSelf().OfType<TExpressionSyntax>()
+                          where NodeMatchesExpression(originalSemanticModel, expressionInOriginal, nodeInCurrent, allOccurrences, cancellationToken)
+                          select nodeInCurrent;
+            return matches;
+        }
+
+        private static bool NodeMatchesExpression(SemanticModel originalSemanticModel,
+            TExpressionSyntax expressionInOriginal, TExpressionSyntax nodeInCurrent, bool allOccurrences,
+            CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+            if (nodeInCurrent == expressionInOriginal)
+            {
+                return true;
+            }
+
+            if (allOccurrences)
```

Comment:
please invert this.  it's much more confusing this way.  it reads much more clearly as "oh, if i'm *not* matching all occurrences, then i don't even need ot proceed".

also, this logic is somewhat suspect.  you are doing the expensive work (walking all expressions) prior to the check if you even need to do the walk at all.  do the check first, and only do the expensive walk if necessayr.

Reply:
inverted order in both